### PR TITLE
Allow FETCH PARTITION from other zookeepers

### PR DESCRIPTION
**Pull request by amosbird**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Now it's possible to `ALTER TABLE table_name FETCH PARTITION partition_expr FROM 'zk://<host>:<port>/path-in-zookeeper'`. It's useful for shipping data to new clusters.

Detailed description / Documentation draft:

Documentation needs syncing.

### CODE REVIEW

**Comment by alesapin** [comment]
`00834_cancel_http_readonly_queries_on_client_close` known flaky test.

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 1]
```diff
@@ -3253,6 +3253,7 @@ bool StorageReplicatedMergeTree::fetchPart(const String & part_name, const Stora
         }
         else
         {
+            part->is_temp = false;
```
Comment (alesapin) [inline src/Storages/StorageReplicatedMergeTree.cpp:3256]: Add a comment on why we need it. It's really important.

**Review by alesapin — APPROVED** [review]
LGTM, but need to add a test at least with our own ZK.

**Comment by alexey-milovidov** [comment]
Does it respect `remote_host_filter`?
Otherwise we have SSRF and the possibility of DoS via ClickHouse.

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 2]
```diff
@@ -4553,9 +4554,17 @@ void StorageReplicatedMergeTree::fetchPartition(
     const String & from_,
     const Context & query_context)
 {
+    Poco::URI uri(from_);
+    String zkhost;
+
+    if (uri.getScheme() == "zk")
```
Comment (alexey-milovidov) [inline src/Storages/StorageReplicatedMergeTree.cpp:4560]: It `zk` somehow standard for URLs?

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 3]
```diff
@@ -4553,9 +4554,17 @@ void StorageReplicatedMergeTree::fetchPartition(
...
+    {
+        zkhost = uri.getHost() + ":" + toString(uri.getPort());
```
Comment (alexey-milovidov) [inline src/Storages/StorageReplicatedMergeTree.cpp:4562]: Looks like IPv6 is not supported?

[::]:123

Comment (alexey-milovidov) [reply src/Storages/StorageReplicatedMergeTree.cpp:4562]: ZooKeeper is a cluster... how to write multiple hosts in URL?

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
Looking for the possiblity of SSRF it will be better to add the support for remote_host_filter and also add a setting in config.xml to enable these queries that is disabled by default (because almost no one enables remote_host_filter).

**Comment by alexey-milovidov** [comment]
The overall solution is limited because it does not support ZK clusters with authentication.
It also looks like it does not allow to connect to ZK with TLS.

Maybe it will be better to have multiple named ZK clusters in config?
It will open up the possibility to specify different ZK clusters for different Replicated tables in future.

**Comment by amosbird** [comment]
@alexey-milovidov reimplemented as suggested.

**Review by alesapin — COMMENTED** [review]
I'm not sure about the name `<auxiliary_zookeepers>` in config.xml if we have any plans to use it for other purposes. The rest part is OK.

**Review by alesapin — APPROVED** [review]
LGTM

**Pull request merged**